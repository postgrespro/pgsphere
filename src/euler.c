#include "euler.h"

/* Euler transformation functions */

PG_FUNCTION_INFO_V1(spheretrans_in);
PG_FUNCTION_INFO_V1(spheretrans_from_float8);
PG_FUNCTION_INFO_V1(spheretrans_from_float8_and_type);
PG_FUNCTION_INFO_V1(spheretrans_equal);
PG_FUNCTION_INFO_V1(spheretrans_not_equal);
PG_FUNCTION_INFO_V1(spheretrans_phi);
PG_FUNCTION_INFO_V1(spheretrans_theta);
PG_FUNCTION_INFO_V1(spheretrans_psi);
PG_FUNCTION_INFO_V1(spheretrans_type);
PG_FUNCTION_INFO_V1(spheretrans_zxz);
PG_FUNCTION_INFO_V1(spheretrans);
PG_FUNCTION_INFO_V1(spheretrans_invert);
PG_FUNCTION_INFO_V1(spheretrans_trans);
PG_FUNCTION_INFO_V1(spheretrans_trans_inv);
PG_FUNCTION_INFO_V1(spheretrans_point);
PG_FUNCTION_INFO_V1(spheretrans_point_inverse);

/*
 * Checks and modifies the Euler transformation.
 */
static void
spheretrans_check(SEuler *e)
{
	SPoint		sp[3];

	sp[0].lat = sp[1].lat = sp[2].lat = 0.0;
	sp[0].lng = e->phi;
	sp[1].lng = e->theta;
	sp[2].lng = e->psi;
	spoint_check(&sp[0]);
	spoint_check(&sp[1]);
	spoint_check(&sp[2]);
	e->phi = sp[0].lng;
	e->theta = sp[1].lng;
	e->psi = sp[2].lng;
}

Datum
spheretrans_in(PG_FUNCTION_ARGS)
{
	SEuler	   *se = (SEuler *) palloc(sizeof(SEuler));
	char	   *c = PG_GETARG_CSTRING(0);
	unsigned char etype[3];
	int			i;

	void		sphere_yyparse(void);

	init_buffer(c);
	sphere_yyparse();

	if (get_euler(&se->phi, &se->theta, &se->psi, etype))
	{

		for (i = 0; i < 3; i++)
		{
			switch (i)
			{
				case 0:
					se->phi_a = etype[i];
					break;
				case 1:
					se->theta_a = etype[i];
					break;
				case 2:
					se->psi_a = etype[i];
					break;
			}
		}
		spheretrans_check(se);
	}
	else
	{
		reset_buffer();
		pfree(se);
		se = NULL;
		elog(ERROR, "spheretrans_in: parse error");
	}
	reset_buffer();
	PG_RETURN_POINTER(se);
}

Datum
spheretrans_from_float8(PG_FUNCTION_ARGS)
{
	SEuler	   *se = (SEuler *) palloc(sizeof(SEuler));

	se->phi = PG_GETARG_FLOAT8(0);
	se->theta = PG_GETARG_FLOAT8(1);
	se->psi = PG_GETARG_FLOAT8(2);
	seuler_set_zxz(se);
	spheretrans_check(se);
	PG_RETURN_POINTER(se);
}

Datum
spheretrans_from_float8_and_type(PG_FUNCTION_ARGS)
{
	SEuler	   *se;
	Datum		d[3];
	int			i;
	char	   *c = PG_GETARG_CSTRING(3);
	unsigned char t = 0;

	d[0] = PG_GETARG_DATUM(0);
	d[1] = PG_GETARG_DATUM(1);
	d[2] = PG_GETARG_DATUM(2);
	se = (SEuler *) DatumGetPointer(DirectFunctionCall3(spheretrans_from_float8,
														d[0], d[1], d[2]));

	for (i = 0; i < 3; i++)
	{
		switch (c[i])
		{
			case 'x':
			case 'X':
				t = EULER_AXIS_X;
				break;
			case 'y':
			case 'Y':
				t = EULER_AXIS_Y;
				break;
			case 'z':
			case 'Z':
				t = EULER_AXIS_Z;
				break;
			default:
				t = 0;
		}

		if (t == 0)
		{
			pfree(se);
			elog(ERROR, "invalid axis format");
		}
		switch (i)
		{
			case 0:
				se->phi_a = t;
				break;
			case 1:
				se->theta_a = t;
				break;
			case 2:
				se->psi_a = t;
				break;
		}
	}
	PG_RETURN_POINTER(se);

}

void
seuler_set_zxz(SEuler *se)
{
	se->phi_a = EULER_AXIS_Z;
	se->theta_a = EULER_AXIS_X;
	se->psi_a = EULER_AXIS_Z;
}

bool
strans_eq(const SEuler *e1, const SEuler *e2)
{
	SPoint		in[2],
				p[4];

	in[0].lng = 0.0;
	in[0].lat = 0.0;
	in[1].lng = PIH;
	in[1].lat = 0.0;

	euler_spoint_trans(&p[0], &in[0], e1);
	euler_spoint_trans(&p[1], &in[1], e1);
	euler_spoint_trans(&p[2], &in[0], e2);
	euler_spoint_trans(&p[3], &in[1], e2);

	return spoint_eq(&p[0], &p[2]) && spoint_eq(&p[1], &p[3]);
}

Datum
spheretrans_equal(PG_FUNCTION_ARGS)
{
	SEuler	   *e1 = (SEuler *) PG_GETARG_POINTER(0);
	SEuler	   *e2 = (SEuler *) PG_GETARG_POINTER(1);

	PG_RETURN_BOOL(strans_eq(e1, e2));
}

Datum
spheretrans_not_equal(PG_FUNCTION_ARGS)
{
	SEuler	   *e1 = (SEuler *) PG_GETARG_POINTER(0);
	SEuler	   *e2 = (SEuler *) PG_GETARG_POINTER(1);

	PG_RETURN_BOOL(!strans_eq(e1, e2));
}

Datum
spheretrans_phi(PG_FUNCTION_ARGS)
{
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(se->phi);
}

Datum
spheretrans_theta(PG_FUNCTION_ARGS)
{
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(se->theta);
}

Datum
spheretrans_psi(PG_FUNCTION_ARGS)
{
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(se->psi);
}

Datum
spheretrans_type(PG_FUNCTION_ARGS)
{
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(0);
	BpChar	   *result = (BpChar *) palloc(3 + VARHDRSZ);
	char		ret[4];
	int			i;
	unsigned char t = 0;

	for (i = 0; i < 3; i++)
	{
		switch (i)
		{
			case 0:
				t = se->phi_a;
				break;
			case 1:
				t = se->theta_a;
				break;
			case 2:
				t = se->psi_a;
				break;
		}
		switch (t)
		{
			case EULER_AXIS_X:
				ret[i] = 'X';
				break;
			case EULER_AXIS_Y:
				ret[i] = 'Y';
				break;
			case EULER_AXIS_Z:
				ret[i] = 'Z';
				break;
		}
	}
	ret[3] = '\0';
	SET_VARSIZE(result, 3 + VARHDRSZ);
	memcpy((void *) VARDATA(result), (void *) &ret[0], 3);

	PG_RETURN_BPCHAR_P(result);
}

void
spheretrans_inv(SEuler *se)
{
	float8		lng[3];
	const unsigned char c = se->phi_a;

	lng[2] = -se->phi;
	lng[1] = -se->theta;
	lng[0] = -se->psi;
	se->phi = lng[0];
	se->theta = lng[1];
	se->psi = lng[2];
	se->phi_a = se->psi_a;
	se->psi_a = c;
}

void
spheretrans_inverse(SEuler *se_out, const SEuler *se_in)
{
	memcpy((void *) se_out, (void *) se_in, sizeof(SEuler));
	spheretrans_inv(se_out);
}

void
strans_zxz(SEuler *ret, const SEuler *se)
{
	if (se->phi_a == EULER_AXIS_Z &&
		se->theta_a == EULER_AXIS_X &&
		se->psi_a == EULER_AXIS_Z)
	{
		memcpy((void *) ret, (void *) se, sizeof(SEuler));
	}
	else
	{
		SEuler		tmp;

		tmp.psi = 0.0;
		tmp.theta = 0.0;
		tmp.phi = 0.0;
		seuler_set_zxz(&tmp);
		seuler_trans_zxz(ret, se, &tmp);
	}
}

Datum
spheretrans_zxz(PG_FUNCTION_ARGS)
{
	SEuler	   *si = (SEuler *) PG_GETARG_POINTER(0);
	SEuler	   *ret = (SEuler *) palloc(sizeof(SEuler));

	strans_zxz(ret, si);
	PG_RETURN_POINTER(ret);
}

Datum
spheretrans(PG_FUNCTION_ARGS)
{
	Datum		d = PG_GETARG_DATUM(0);

	PG_RETURN_DATUM(d);
}

Datum
spheretrans_invert(PG_FUNCTION_ARGS)
{
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(0);
	SEuler	   *ret = (SEuler *) palloc(sizeof(SEuler));

	spheretrans_inverse(ret, se);
	PG_RETURN_POINTER(ret);
}

void
seuler_trans_zxz(SEuler *out, const SEuler *in, const SEuler *se)
{
	SPoint		sp[4];

	sp[0].lng = 0.0;
	sp[0].lat = 0.0;
	sp[1].lng = PIH;
	sp[1].lat = 0.0;
	euler_spoint_trans(&sp[2], &sp[0], in);
	euler_spoint_trans(&sp[3], &sp[1], in);
	euler_spoint_trans(&sp[0], &sp[2], se);
	euler_spoint_trans(&sp[1], &sp[3], se);
	spherevector_to_euler(out, &sp[0], &sp[1]);
}

Datum
spheretrans_trans(PG_FUNCTION_ARGS)
{
	SEuler	   *se1 = (SEuler *) PG_GETARG_POINTER(0);
	SEuler	   *se2 = (SEuler *) PG_GETARG_POINTER(1);
	SEuler	   *out = (SEuler *) palloc(sizeof(SEuler));

	seuler_trans_zxz(out, se1, se2);
	PG_RETURN_POINTER(out);
}

Datum
spheretrans_trans_inv(PG_FUNCTION_ARGS)
{
	SEuler	   *se1 = (SEuler *) PG_GETARG_POINTER(0);
	SEuler	   *se2 = (SEuler *) PG_GETARG_POINTER(1);
	SEuler	   *out = (SEuler *) palloc(sizeof(SEuler));
	SEuler		tmp;

	spheretrans_inverse(&tmp, se2);
	seuler_trans_zxz(out, se1, &tmp);
	spheretrans_check(out);
	PG_RETURN_POINTER(out);
}

void
euler_spoint_trans(SPoint *out, const SPoint *in, const SEuler *se)
{
	Vector3D	v,
				o;

	spoint_vector3d(&v, in);
	euler_vector_trans(&o, &v, se);
	vector3d_spoint(out, &o);
}

Datum
spheretrans_point(PG_FUNCTION_ARGS)
{
	SPoint	   *sp = (SPoint *) PG_GETARG_POINTER(0);
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(1);
	SPoint	   *out = (SPoint *) palloc(sizeof(SPoint));

	euler_spoint_trans(out, sp, se);
	PG_RETURN_POINTER(out);
}

Datum
spheretrans_point_inverse(PG_FUNCTION_ARGS)
{
	Datum		sp = PG_GETARG_DATUM(0);
	SEuler	   *se = (SEuler *) PG_GETARG_POINTER(1);
	SEuler		tmp;
	Datum		ret;

	spheretrans_inverse(&tmp, se);
	ret = DirectFunctionCall2(spheretrans_point,
							  sp, PointerGetDatum(&tmp));
	PG_RETURN_DATUM(ret);
}

 /*
  * Transforms a spherical vector from 'spb' to 'spe' into an inverse Euler
  * transformation. Returns true if the transformation was successful.
  */
static bool
spherevector_to_euler_inv(SEuler *se, const SPoint *spb, const SPoint *spe)
{
	if (spoint_eq(spb, spe))
	{
		return false;
	}
	else
	{
		Vector3D	vbeg,
					vend,
					vtmp;
		SPoint		spt[2];
		SEuler		set;

		spoint_vector3d(&vbeg, spb);
		spoint_vector3d(&vend, spe);
		vector3d_cross(&vtmp, &vbeg, &vend);
		vector3d_spoint(&spt[0], &vtmp);
		set.phi = -spt[0].lng - PIH;
		set.theta = spt[0].lat - PIH;
		set.psi = 0.0;
		seuler_set_zxz(&set);
		euler_spoint_trans(&spt[1], spb, &set);
		set.psi = -spt[1].lng;
		memcpy((void *) se, (void *) &set, sizeof(SEuler));
	}

	return true;
}

bool
spherevector_to_euler(SEuler *se, const SPoint *spb, const SPoint *spe)
{
	bool		ret;

	ret = spherevector_to_euler_inv(se, spb, spe);
	if (ret)
	{
		spheretrans_inv(se);
	}
	return ret;
}

void
euler_vector_trans(Vector3D *out, const Vector3D *in, const SEuler *se)
{
	int			i;
	unsigned char t;
	const double *a;
	double		u[3],
				vr[3],
				sa,
				ca;

	t = 0;
	a = NULL;
	u[0] = in->x;
	u[1] = in->y;
	u[2] = in->z;

	for (i = 0; i < 3; i++)
	{

		switch (i)
		{
			case 0:
				a = &se->phi;
				t = se->phi_a;
				break;
			case 1:
				a = &se->theta;
				t = se->theta_a;
				break;
			case 2:
				a = &se->psi;
				t = se->psi_a;
				break;
		}

		if (FPzero(*a))
		{
			continue;
		}

		sa = sin(*a);
		ca = cos(*a);

		switch (t)
		{

			case EULER_AXIS_X:
				vr[0] = u[0];
				vr[1] = ca * u[1] - sa * u[2];
				vr[2] = sa * u[1] + ca * u[2];
				break;
			case EULER_AXIS_Y:
				vr[0] = ca * u[0] + sa * u[2];
				vr[1] = u[1];
				vr[2] = ca * u[2] - sa * u[0];
				break;
			case EULER_AXIS_Z:
				vr[0] = ca * u[0] - sa * u[1];
				vr[1] = sa * u[0] + ca * u[1];
				vr[2] = u[2];
				break;

		}
		memcpy((void *) &u[0], (void *) &vr[0], sizeof(u));

	}
	out->x = u[0];
	out->y = u[1];
	out->z = u[2];
}
