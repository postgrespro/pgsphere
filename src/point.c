#include "point.h"

/* This file contains definitions for spherical point functions. */

PG_FUNCTION_INFO_V1(spherepoint_in);
PG_FUNCTION_INFO_V1(spherepoint_from_long_lat);
PG_FUNCTION_INFO_V1(spherepoint_distance);
PG_FUNCTION_INFO_V1(spherepoint_long);
PG_FUNCTION_INFO_V1(spherepoint_lat);
PG_FUNCTION_INFO_V1(spherepoint_x);
PG_FUNCTION_INFO_V1(spherepoint_y);
PG_FUNCTION_INFO_V1(spherepoint_z);
PG_FUNCTION_INFO_V1(spherepoint_xyz);
PG_FUNCTION_INFO_V1(spherepoint_equal);
PG_FUNCTION_INFO_V1(centroid);

bool
spoint_eq(const SPoint *p1, const SPoint *p2)
{
	Vector3D	a,
				b;

	spoint_vector3d(&a, p1);
	spoint_vector3d(&b, p2);
	return (vector3d_eq(&a, &b));
}

void
spoint_check(SPoint *spoint)
{
	bool lat_is_neg = (spoint->lat < 0) ? true : false;

	if (spoint->lng < 0 || spoint->lng > PID)
		spoint->lng = spoint->lng - floor(spoint->lng / (PID)) * PID;
	if (spoint->lat < -PIH || spoint->lat > PIH)
		spoint->lat = spoint->lat - floor(spoint->lat / (PID)) * PID;
	if (spoint->lng < 0.0)
	{
		spoint->lng += (PID);
	}
	if (spoint->lat > PI)
	{
		spoint->lat -= (2 * PI);
	}
	if (spoint->lat > PIH)
	{
		spoint->lat = (PI - spoint->lat);
		spoint->lng += ((spoint->lng < PI) ? (PI) : (-PI));
	}
	if (spoint->lat < -PIH)
	{
		spoint->lat = (-PI - spoint->lat);
		spoint->lng += ((spoint->lng < PI) ? (PI) : (-PI));
	}
	if (FPeq(spoint->lat, PIH) && lat_is_neg)
		spoint->lat = -PIH;

	if (FPeq(spoint->lng, PID))
	{
		spoint->lng = 0.0;
	}
	if (FPzero(spoint->lng))
	{
		spoint->lng = 0.0;
	}
	if (FPzero(spoint->lat))
	{
		spoint->lat = 0.0;
	}
}

void
vector3d_spoint(SPoint *p, const Vector3D *v)
{
	double rho = sqrt((v->x) * (v->x) + (v->y) * (v->y));

	if (0.0 == rho)
	{
		if (FPzero(v->z))
		{
			p->lat = 0.0;
		}
		else if (v->z > 0)
		{
			p->lat = PIH;
		}
		else if (v->z < 0)
		{
			p->lat = -PIH;
		}
	}
	else
	{
		p->lat = atan(v->z / rho);
	}

	p->lng = atan2(v->y, v->x);
	if (FPzero(p->lng))
	{
		p->lng = 0.0;
	}
	else if (p->lng < 0.0)
	{
		p->lng += PID;
	}
}

void
spoint_vector3d(Vector3D *v, const SPoint *p)
{
	v->x = cos(p->lng) * cos(p->lat);
	v->y = sin(p->lng) * cos(p->lat);
	v->z = sin(p->lat);
}

Datum
spherepoint_in(PG_FUNCTION_ARGS)
{
	SPoint	   *sp = (SPoint *) palloc(sizeof(SPoint));
	char	   *c = PG_GETARG_CSTRING(0);
	double		lng,
				lat;

	void		sphere_yyparse(void);

	init_buffer(c);
	sphere_yyparse();
	if (get_point(&lng, &lat))
	{
		sp->lng = lng;
		sp->lat = lat;
		spoint_check(sp);
	}
	else
	{
		reset_buffer();
		pfree(sp);
		sp = NULL;
		elog(ERROR, "spherepoint_in: parse error");
	}
	reset_buffer();
	PG_RETURN_POINTER(sp);
}


Datum
spherepoint_from_long_lat(PG_FUNCTION_ARGS)
{
	SPoint *p = (SPoint *) palloc(sizeof(SPoint));

	p->lng = PG_GETARG_FLOAT8(0);
	p->lat = PG_GETARG_FLOAT8(1);
	spoint_check(p);
	PG_RETURN_POINTER(p);
}

static double
norm2(double a, double b)
{
	return sqrt(a * a + b * b);
}

float8
spoint_dist(const SPoint *p1, const SPoint *p2)
{
	float8	dl = p1->lng - p2->lng;
	/* use Vincenty's formula for the inverse geodesic problem on the sphere */
	float8	f = atan2(norm2(cos(p2->lat) * sin(dl),
							cos(p1->lat) * sin(p2->lat)
								- sin(p1->lat) * cos(p2->lat) * cos(dl)),
						sin(p1->lat) * sin(p2->lat)
							+ cos(p1->lat) * cos(p2->lat) * cos(dl));
	if (FPzero(f))
	{
		return 0.0;
	}
	else
	{
		return f;
	}
}

Datum
spherepoint_distance(PG_FUNCTION_ARGS)
{
	SPoint	   *p1 = (SPoint *) PG_GETARG_POINTER(0);
	SPoint	   *p2 = (SPoint *) PG_GETARG_POINTER(1);

	PG_RETURN_FLOAT8(spoint_dist(p1, p2));

}

Datum
spherepoint_long(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(p->lng);
}

Datum
spherepoint_lat(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(p->lat);
}

Datum
spherepoint_x(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	Vector3D	v;

	spoint_vector3d(&v, p);
	PG_RETURN_FLOAT8(v.x);
}

Datum
spherepoint_y(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	Vector3D	v;

	spoint_vector3d(&v, p);
	PG_RETURN_FLOAT8(v.y);
}

Datum
spherepoint_z(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	Vector3D	v;

	spoint_vector3d(&v, p);
	PG_RETURN_FLOAT8(v.z);
}

Datum
spherepoint_xyz(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	Datum		dret[3];
	ArrayType  *result;
	Vector3D	v;

	spoint_vector3d(&v, p);
	dret[0] = Float8GetDatumFast(v.x);
	dret[1] = Float8GetDatumFast(v.y);
	dret[2] = Float8GetDatumFast(v.z);
#ifdef USE_FLOAT8_BYVAL
	result = construct_array(dret, 3, FLOAT8OID, sizeof(float8), true, 'd');
#else
	result = construct_array(dret, 3, FLOAT8OID, sizeof(float8), false, 'd');
#endif
	PG_RETURN_ARRAYTYPE_P(result);
}

Datum
spherepoint_equal(PG_FUNCTION_ARGS)
{
	SPoint	   *p1 = (SPoint *) PG_GETARG_POINTER(0);
	SPoint	   *p2 = (SPoint *) PG_GETARG_POINTER(1);

	PG_RETURN_BOOL(spoint_eq(p1, p2));
}

static SPoint * spherepoint_from_vector3d(Vector3D v){
	SPoint* p = (SPoint *) palloc(sizeof(SPoint));
	p->lat = asin(v.z / sqrt(pow(v.x, 2) + pow(v.y, 2)  + pow(v.z, 2)));
	p->lng = atan2(v.y, v.x);
	return p;
}

Datum centroid(PG_FUNCTION_ARGS) {
	int i;
	SPoint * p;
	SPoint current_point;
	Vector3D	v;
	Vector3D	point_coords;
	ArrayType *dots_vector = PG_GETARG_ARRAYTYPE_P(0);
	int num_elements = ArrayGetNItems(ARR_NDIM(dots_vector), ARR_DIMS(dots_vector));
	SPoint *array_data = (SPoint *) ARR_DATA_PTR(dots_vector);

	for (i = 0; i < num_elements; i++) {
		current_point = array_data[i];
		spoint_vector3d(&v, &current_point);
		point_coords.x+=v.x;
		point_coords.y+=v.y;
		point_coords.z+=v.z;
	}

	point_coords.x /= num_elements;
	point_coords.y /= num_elements;
	point_coords.z /= num_elements;

	p = spherepoint_from_vector3d(point_coords);

	spoint_check(p);

	PG_RETURN_POINTER(p);
}
