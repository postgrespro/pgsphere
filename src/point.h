#ifndef __PGS_POINT_H__
#define __PGS_POINT_H__

#include "vector3d.h"
#include "sbuffer.h"

/* This file contains declarations for spherical point and functions. */

/*
 * The data structure definition of a spherical point.
 */
typedef struct
{
	float8		lng;			/* longitude value in radians */
	float8		lat;			/* latitude value in radians */
} SPoint;

extern Oid	get_spoint_type_oid(void);

/*
 * Calculate the distance between two spherical points in radians.
 */

extern float8 spoint_dist(const SPoint *p1, const SPoint *p2);

/*
 * Check whether two points are equal.
 */

extern bool spoint_eq(const SPoint *p1, const SPoint *p2);

/*
 * Check the longitude and latitude values of a spherical point.
 */

extern void spoint_check(SPoint *spoint);

/*
 * Transforms a 3d vector into a spherical point.
 */

extern void vector3d_spoint(SPoint *p, const Vector3D *v);

/*
 * Transforms a spherical point into a 3d vector.
 */

extern void spoint_vector3d(Vector3D *v, const SPoint *p);

/*
 * Take the input and store it as a spherical point.
 */

extern Datum spherepoint_in(PG_FUNCTION_ARGS);

/*
 * Create spherical point from lat, lng and store to first argument(pointer)
 */

extern void create_spherepoint_from_long_lat(SPoint *p, float8 lng, float8 lat);

/*
 * Create a spherical point from longitude and latitude both in radians.
 */

extern Datum spherepoint_from_long_lat(PG_FUNCTION_ARGS);

/*
 * Create a spherical point from longitude and latitude both in degrees.
 */

extern Datum spherepoint_from_long_lat_deg(PG_FUNCTION_ARGS);

/*
 * Calculate the distance between two spherical points.
 */
extern Datum spherepoint_distance(PG_FUNCTION_ARGS);

/*
 * Longitude of a spherical point.
 */
extern Datum spherepoint_long(PG_FUNCTION_ARGS);

/*
 * Latitude of a spherical point.
 */
extern Datum spherepoint_lat(PG_FUNCTION_ARGS);

/*
 * Cartesian x-value of a spherical point.
 */
extern Datum spherepoint_x(PG_FUNCTION_ARGS);

/*
 * Cartesian y-value of a spherical point.
 */
extern Datum spherepoint_y(PG_FUNCTION_ARGS);

/*
 * Cartesian z-value of a spherical point.
 */
extern Datum spherepoint_z(PG_FUNCTION_ARGS);

/*
 * Cartesian values of a spherical point as an array.
 */
extern Datum spherepoint_xyz(PG_FUNCTION_ARGS);

/*
 * Check whether two points are equal.
 */
extern Datum spherepoint_equal(PG_FUNCTION_ARGS);

/*
 * Compute a 32-bit hash value of a point.
 */
extern Datum spherepoint_hash32(PG_FUNCTION_ARGS);

#endif
