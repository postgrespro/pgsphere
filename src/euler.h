#ifndef __PGS_EULER_H__
#define __PGS_EULER_H__

#include "point.h"

/*
 * Euler transformation declarations
 */

/*
 * Data structure of spherical (Euler) transformation.
 */
typedef struct
{
	unsigned char	phi_a:2,	/* first axis */
					theta_a:2,	/* second axis */
					psi_a:2;	/* third axis */
	float8			phi,		/* first rotation angle */
					theta,		/* second rotation angle */
					psi;		/* third rotation angle */
} SEuler;


/*
 * Transforms a spherical point and returns the pointer to a transformed spherical
 * point.
 */
void	euler_spoint_trans(SPoint *out, const SPoint *in, const SEuler *se);

/*
 * Transforms a spherical vector from 'spb' to 'spe' into an Euler transformation.
 * Returns true if the transformation was successful.
 */
bool	spherevector_to_euler(SEuler *se, const SPoint *spb, const SPoint *spe);

/*
 * Sets the axes of transformation to ZXZ.
 */
void	seuler_set_zxz(SEuler *se);

/*
 * Checks equality of two transformations.
 */
bool	strans_eq(const SEuler *e1, const SEuler *e2);

/*
 * Transforms a vector using an Euler transformation. Returns the pointer to
 * the result vector.
 */
void	euler_vector_trans(Vector3D *out, const Vector3D *in, const SEuler *se);

/*
 * Inverts an Euler transformation. Returns the pointer to the
 * inverted transformation.
 */
void	spheretrans_inverse(SEuler *se_out, const SEuler *se_in);

/*
 * Inverts an Euler transformation replacing the original Euler transformation.
 * Returns the pointer to the inverted transformation.
 */
void	spheretrans_inv(SEuler *se);

/*
 * Converts an Euler transformation to a ZXZ-axis transformation. Returns
 * the pointer to the converted transformation.
 */
void	strans_zxz(SEuler *ret, const SEuler *se);

/*
 * Transforms an Euler transformation 'in' into 'out' using 'se'. The result
 * is always a ZXZ-axis transformation. Returns the pointer to the transformed
 * transformation.
 */
void	seuler_trans_zxz(SEuler *out, const SEuler *in, const SEuler *se);

/*
 * Input of an Euler transformation.
 */
Datum	spheretrans_in(PG_FUNCTION_ARGS);

/*
 * Input of an Euler transformation with axis Z,X,Z from three angles
 * (phi, theta, psi) in radians.
 */
Datum	spheretrans_from_float8(PG_FUNCTION_ARGS);

/*
 * Returns the first angle of an Euler transformation in radians.
 */
Datum	spheretrans_phi(PG_FUNCTION_ARGS);

/*
 * Returns the second angle of an Euler transformation in radians.
 */
Datum	spheretrans_theta(PG_FUNCTION_ARGS);

/*
 * Returns the third angle of an Euler transformation in radians.
 */
Datum	spheretrans_psi(PG_FUNCTION_ARGS);

/*
 * Returns the axis of an Euler transformation as three letter code.
 */
Datum	spheretrans_type(PG_FUNCTION_ARGS);

/*
 * Returns the Euler transformation (does nothing). This function is needed
 * for +strans operator.
 */
Datum	spheretrans(PG_FUNCTION_ARGS);

/*
 * Returns the inverse Euler transformation.
 */
Datum	spheretrans_invert(PG_FUNCTION_ARGS);

/*
 * Convert an Euler transformation to a ZXZ-axis transformation.
 */
Datum	spheretrans_zxz(PG_FUNCTION_ARGS);

/*
 * This function creates an Euler transformation from 3 angle values in
 * radians and three letter code used for axes. A letter can be X, Y or Z
 * (case-insensitive).
 */
Datum	spheretrans_from_float8_and_type(PG_FUNCTION_ARGS);

/*
 * Checks equality of two Euler transformations.
 */
Datum	spheretrans_equal(PG_FUNCTION_ARGS);

/*
 * Checks inequality of two Euler transformations.
 */
Datum	spheretrans_not_equal(PG_FUNCTION_ARGS);

/*
 * Transforms an Euler transformation.
 */
Datum	spheretrans_trans(PG_FUNCTION_ARGS);

/*
 * Transforms inverse an Euler transformations.
 */
Datum	spheretrans_trans_inv(PG_FUNCTION_ARGS);

/*
 * Transforms a spherical point.
 */
Datum	spheretrans_point(PG_FUNCTION_ARGS);

/*
 * Perform inverse transformation of a spherical point.
 */
Datum	spheretrans_point_inverse(PG_FUNCTION_ARGS);

#endif
