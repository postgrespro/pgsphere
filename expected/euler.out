\set ECHO none
 set_sphere_output_precision 
-----------------------------
 SET 8
(1 row)

-- checking Euler transformation operators
SELECT strans '-10d,0d,10d,ZZZ'  = '-10d,0d,10d,XXX' ;
 ?column? 
----------
 t
(1 row)

SELECT strans '-40d,0d,40d,ZZZ' <> '-40d,0d,40d,XXX' ;
 ?column? 
----------
 f
(1 row)

